File: twilio/orchestration/java/cloud-infrastructure-model/src/main/java/com/twilio/interview/cloudinfrastructure/model/GroupType.java
package com.twilio.interview.cloudinfrastructure.model;

import java.util.List;

/**
 * Interface {@link GroupType} defines the contract implemented by the host
 * group type objects.
 */
public interface GroupType extends HostType {
    /**
     * Returns host (instance) size for the group.
     * 
     * @return  host size
     */
    HostSize getSize ();

    /**
     * Returns the number of hosts (instances) for the group.
     * 
     * @return  host count
     */
    int getHostCount ();

    /**
     * Returns the flag indicating whether host group should be activated
     * (brought into the load balancer) after successful boot or not.
     * 
     * @return  activate on boot flag
     */
    boolean isActiveOnBoot ();

    /**
     * Returns the list of {@link GroupType}s the group depends on.
     * 
     * @return  group dependencies
     */
    List<GroupType> getDependencies ();
}

File: twilio/orchestration/php/lib/GroupType.php
<?php
 
class GroupType {
	protected $size;
	protected $count;
	protected $activateOnBoot;
	protected $dependencies;

	public function __construct($size=Size::SMALL, $count=1, $activateOnBoot=false, $dependencies=[]) {
		$this->size = $size;
		$this->count = $count;
		$this->activateOnBoot = $activateOnBoot;
		$this->dependencies = $dependencies;
	}

	/**
	 * Get the ActivateOnBoot configuration setting for this GroupType
	 * @return bool Whether or not the hosts should activate once running
	 */
	public function getActivateOnBoot() {
		return $this->activateOnBoot;
	}

	/**
	 * Get the number of hosts in the GroupType
	 * @return int Number of hosts in the GroupType
	 */
	public function getCount() {
		return $this->count;
	}

	/**
	 * Get the array of GroupTypes the group depends on
	 * @return GroupType[] Dependencies array
	 */
	public function getDependencies() {
		return $this->dependencies;
	}

	/**
	 * Get the size of hosts in the GroupType
	 * @return string Size of Host to boot
	 */
	public function getSize() {
		return $this->size;
	}


}
